proof/refine/ARM/Reply_R.thy

```diff
@@ -41,25 +38,30 @@ lemma replyUnlink_st_tcb_at'_sym_ref:
   apply (wpsimp wp: replyUnlink_st_tcb_at'_wp)
   apply (clarsimp simp: obj_at'_def)
   done
+*)
 
-lemma replyRemoveTCB_st_tcb_at'_Inactive:
-  "\<lbrace>\<lambda>s. tcb_at' tptr s \<longrightarrow> st_tcb_at' test tptr s \<and> test Inactive\<rbrace>
+lemma replyRemoveTCB_st_tcb_at'_Inactive':
+  "\<lbrace>\<top>\<rbrace>
    replyRemoveTCB tptr
-   \<lbrace>\<lambda>_. st_tcb_at' test tptr\<rbrace>"
+   \<lbrace>\<lambda>_. st_tcb_at' ((=) Inactive) tptr\<rbrace>"
```

Comment:
Isn't this going to cause issues with goals like `st_tcb_at' (Not o runnable')` or whatever?

Reply:
Agreed, probably good to leave the more general result in there.

Reply:
Out of curiosity, in what sense is it fiddly?

Reply:
I don’t quite remember, sorry. It could have been just being non-trivial, but we don’t really need this form anyway, at least at the moment. And I am not sure if the original lemma is actually more general, honestly.

Reply:
We could keep the general form, but instantiate and `[wp]` the `Inactive` form with `lemmas`

Reply:
But the preconditions are different. ..

Reply:
> In the old theorem, I suspect the st_tcb_at' part of the precondition is not necessary. Perhaps the best of both worlds might be to keep the test, but remove the st_tcb_at' from the precondition.

I think that is correct, yes.

Reply:
Well, the lemma I wrote can’t be written as an instantiation of the original one. There is something a bit odd about the original lemma, actually. `replyUnlink t` sets the thread state of `t` to `Inactive` at the end. And `t` is in `BlockedOnReply` at the beginning of the function. So, the original lemma is true only for `test` that happens to be true for both `BlockedOnReply` and `Inactive`. So, it cannot be instantiated to `(=) Inactive`.

This original one is not really general because we do want all `P` that holds for `Inactive` only in the conclusion, regardless of the precondition (without being constrained by `BlockedOnReply`).

And this is only used in
```
lemma cancelIPC_simple[wp]:
   "\<lbrace>\<top>\<rbrace> cancelIPC t \<lbrace>\<lambda>rv. st_tcb_at' simple' t\<rbrace>”
```
which happens to have `simple’` as `test`.

Reply:
And, really, the `wp` rule Matt is talking about is just the same thing as combining the lemma I wrote with `pred_tcb’_weakenE`.